### ISSUE TITLE
Better fromListWith

### ISSUE THREAD

**Issue by andrewthad** [tibbe/unordered-containers#162]
During the resolution of a different issue, @treeowl made a [branch with a different definition of fromListWith](https://github.com/treeowl/unordered-containers/tree/super-safe). There are two advantages of this approach.

1. It's easier to be confident that it behaves correctly. David's claim is that the "unsafe modification functions that modify arrays in place are utterly bogus". It unclear whether or not anyone will be able to make the potential unsoundness manifest itself, but it's certainly easier to look at David's alternative implementation and feel confident that there's no unintended aliasing going on.
2. Concerning performance, the `fromListWith` in the branch may generate less garbage and perform slightly better than the existing one. Since the existing `fromListWith` relies on the repeated freezing and thawing of arrays, it's possible that they could be eagerly promoted (copied by the GC), only to later be thawed, mutated, frozen and eagerly promoted again. The [GHC wiki](https://ghc.haskell.org/trac/ghc/wiki/Commentary/Rts/Storage/GC/EagerPromotion) claims that mutable arrays never get eagerly promoted until they are frozen, so the new implementation may benefit from avoiding unnecessary eager promotions. I've not benchmarked this, so it could turn out to be a negligible gain.

**Comment by treeowl** [comment]
I think @bgamari had an idea for a way to let GHC freeze them all at once at the end. That would let us use the same algorithm legitimately.

**Comment by andrewthad** [comment]
Ah, I had forgotten about the issue of freezing nested arrays. Yeah, your implementation does suffer from having to duplication everything at the end. Kind of a bummer. Is Ben's idea documented anywhere?

Also, just for clarification, when you say

> That would let us use the same algorithm legitimately.

I'm interpreting this to mean that Ben's idea would allow your implementation to an in-place freeze at the end, giving it identical performance to the existing implementation. Is that correct?

**Comment by treeowl** [comment]
Ben hasn't told me just what his idea is, but I think one way or another
its about tying a bunch of mutable arrays together so that pressing one
button can freeze them all at once. Then I guess you'd unsafeCoerce the
whole tree of them to convert the mutable types to immutable ones, but I'm
not entirely clear on that bit.

**Comment by bgamari** [comment]
I think the idea that @dolio and I threw around went something like this:

Introduce a sort of mutable token that would represent a group of objects on the mutable list,
```haskell
data MutToken#

newMutToken# :: ST  s MutToken#
```
Provide a way to create a mutable array with this token,
```haskell
newMutArray'# :: MutToken# -> Int -> ST s (MutArray# a)
```

Along with a function to freeze the whole bunch. The problem, however, is that it would be very difficult to type the freeze function as it changes the type of a bunch of arrays "at a distance". A believe this can be accomplished via a phantom type parameter (attached to each array identifying whether it is frozen or not) and some funky quantification (with the freeze function providing evidence that the phantom is frozen). However, I suspect this would be terrible to use in practice and this would require reworking most of our array primops (or introducing a new set).

**Comment by treeowl** [comment]
Wouldn't a primitive that doesn't bother with the types (so users would
have to use corrections to some extent) at least be better than nothing? We
could avoid some unsafe coercions in Core by offering something like

unsafeArrayCoercion :: Coercion Array# MutableArray#

Completely unsafe, of course, but better than what unordered-containers has
to do now.

On Jul 12, 2017 1:13 PM, "Ben Gamari" <notifications@github.com> wrote:

I think the idea that @dolio <https://github.com/dolio> and I threw around
went something like this:

Introduce a sort of mutable token that would represent a group of objects
on the mutable list,

data MutToken#

newMutToken# :: ST  s MutToken#

Provide a way to create a mutable array with this token,

newMutArray'# :: MutToken# -> Int -> ST s (MutArray# a)

Along with a function to freeze the whole bunch. The problem, however, is
that it would be very difficult to type the freeze function as it changes
the type of a bunch of arrays "at a distance". A believe this can be
accomplished via a phantom type parameter (attached to each array
identifying whether it is frozen or not) and some funky quantification
(with the freeze function providing evidence that the phantom is frozen).
However, I suspect this would be terrible to use in practice and this would
require reworking most of our array primops (or introducing a new set).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tibbe/unordered-containers/issues/162#issuecomment-314836109>,
or mute the thread

**Comment by andrewthad** [comment]
@treeowl I don't think the `unsafeArrayCoercion` you suggest would have the right behavior. Since `coerceWith` is implemented as a noop:

    coerceWith :: Coercion a b -> a -> b
    coerceWith Coercion x = coerce x

You don't end up properly changing the [closure type](https://ghc.haskell.org/trac/ghc/browser/includes/rts/storage/ClosureTypes.h) like the freeze functions do. And then the array is stuck on the mutable list forever. I am also unable to think of a way to bring Ben's idea under the discipline of types, but the idea of an extremely unsafe api doesn't seem that terrible to me:

    data MutToken# s
    newMutToken# :: ST  s (MutToken# s)
    newMutArray'# :: MutToken# s -> Int -> ST s (MutArray# s a)
    -- after calling freezeToken, all of your types are now a lie, and you
    -- probably need to unsafeCoerce some big data structure
    freezeToken# :: MutToken# s -> ST s ()

**Comment by andrewthad** [comment]
@treeowl Actually, I just realized that I misunderstood what you were trying to communicate. I think we're probably actually saying about the same thing.

**Comment by treeowl** [comment]
Yeah. The best way I see with our present type system is something like

data HashMapF k v (arrHMkv :: TYPE 'UnliftedRep) (arrLkv :: TYPE
'UnliftedRep)
    = Empty
    | BitmapIndexed !Word arrHMkv
    | Leaf !Hash !(Leaf k v)
    | Full arrHMkv
    | Collision !Hash arrLkv

newtype HashMap k v = HashMap (HashMapF k v (Array# (HashMap k v)) (Array#
(Leaf k v)))

newtype MutHashMap s k v = MutHashMap (HashMapF k v (MutableArray# s
(MutHashMap s k v)) (MutableArray# s (Leaf k v)))

foo :: Coercion (MutHashMap k v) (HashMap s k v)
foo = unsafeCoerce (Coercion :: Coercion Any Any)

Once you hit the button to freeze all the arrays in the MutHashMap (GHC
will not check that you got them all!), you can use foo to coerce the whole
thing to a HashMap. Ugly as sin.

**Comment by andrewthad** [comment]
I had actually been thinking about the same thing with the `UnliftedRep`. Ugly, but at least it's easy to verify that it's correct. The real problem is that it won't build on anything before GHC 8.0. Well, and that you'd have to rewrite almost everything in `unordered-containers` since you'd be dealing with the unlifted array type instead of the boxed wrapper. Is it possible to `unsafeCoerce` between identically defined ADTs or is that madness? Like this:

    data Foo = FooA Int Bool | FooB Char
    data Bar = BarA Int Bool | BarB Char

Can we go from `Foo` to `Bar` with `unsafeCoerce`? The `unsafeCoerce#` docs suggest that this is not supported, but if actually is, it seems like a solution.

**Comment by treeowl** [comment]
I think it'd probably work, but it's definitely not supported. Might still
be better than what we do now, though. I think some pattern synonyms could
probably avoid most rewriting of the UnliftedRep version. But all of this
is pretty much moot unless/until we get the multifreeze primop.

On Jul 12, 2017 5:20 PM, "Andrew Martin" <notifications@github.com> wrote:

I had actually been thinking about the same thing with the UnliftedRep.
Ugly, but at least it's easy to verify that it's correct. The real problem
is that it won't build on anything before GHC 8.0. Well, and that you'd
have to rewrite almost everything in unordered-containers since you'd be
dealing with the unlifted array type instead of the boxed wrapper. Is it
possible to unsafeCoerce between identically defined ADTs or is that
madness? Like this:

data Foo = FooA Int Bool | FooB Char
data Bar = BarA Int Bool | BarB Char

Can we go from Foo to Bar with unsafeCoerce? The unsafeCoerce# docs suggest
that this is not supported, but if actually is, it seems like a solution.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tibbe/unordered-containers/issues/162#issuecomment-314900614>,
or mute the thread

**Comment by treeowl** [comment]
@andrewthad, the pattern synonyms for `HashMap` would look like

```haskell
pattern Empty :: HashMap k v
pattern Empty = HashMap EmptyF

pattern BitmapIndexed :: Bitmap -> Array (HashMap k v) -> HashMap k v
pattern BitmapIndexed bm ar <- ((\(HashMap (BitmapIndexedF bm' ar')) -> (# bm', Array ar' #)) -> (# bm, ar #))
  where BitmapIndexed bm (Array ar) = HashMap (BitmapIndexedF bm ar)

pattern Leaf :: Hash -> Leaf k v -> HashMap k v
pattern Leaf h l = HashMap (LeafF h l)

pattern Full :: Array (HashMap k v) -> HashMap k v
pattern Full ar <- ((\(HashMap (FullF ar')) -> Array ar') -> ar)
  where Full (Array ar) = HashMap (FullF ar)

pattern Collision :: Hash -> Array (HashMap k v) -> HashMap k v
pattern Collision h ar <- ((\(HashMap (BitmapIndexedF h' ar')) -> (# h', Array ar' #)) -> (# h, ar #))
  where Collision h (Array ar) = HashMap (BitmapIndexedF h ar)

{-# COMPLETE Empty, BitmapIndexed, Leaf, Full, Collision #-}
```

The ones for `MutHashMap` would be similar.

**Comment by dolio** [comment]
The 'obvious' approach to build a somewhat nicer (but less general) interface to this stuff is to have a scope that allocates the token and freezes it at the end. Everything that gets allocated would be associated to the token by a shared type. It might also be possible to share the token index with the `ST` index.

It would probably help if arrays and mutable arrays were not presented as completely separate types, too. So you could have e.g.:

    Array Immutable t
    Array (Mutable s) t

etc. So you no longer need to parameterize on separate array types, just the mutability status. And the analogue of `runST` that freezes things can have a type like `(forall s. ST s (f (Mutable s))) -> f Immutable` or something.

Arrays are the easy example, though. The tough example is when you want to freeze a data type containing something like `MutVar# s T` into a data type containing `T` instead (possibly with `T` being parameterized by mutability).

**Comment by treeowl** [comment]
@dolio, the obvious barrier to doing this entirely safely is roles. I could write

```haskell
data Foo i where
  I :: Array Immutable () -> Foo Immutable
  M :: Array (Mutable s) () -> Foo (Mutable s)
```

Then `runMutable (blah :: forall s. ST s (Foo (Mutable s)))` produces an `M` constructor with an (operationally) immutable array in it. Disaster! So you'd need to make sure to only allow this when `f`'s parameter is representational, which I *think* lands us in a similarly difficult situation WRT recursive types.

**Comment by andrewthad** [comment]
@dolio That's clever. For a type constructor `f` that does GADT-like analysis of it's arguments, it let's you write `unsafeCoerce`:

    deepST :: (forall s. ST s (f (Mutable s))) -> f Immutable
    
    type family Lower x where
      Lower Immutable = Int
      Lower (Mutable s) = Bool
    
    data Lowered x = Lowered (Lower x)
    
    foo :: Int
    foo = deepST (return (Lowered False :: Lowered (Mutable s)))

Admittedly, this doesn't actually seem like that big of a deal, given that people should only use `deepST` on types they define in a library, not types that end users pass in. It's sort of in the same boat as [unsafeVacuous](http://hackage.haskell.org/package/void-0.7.2/docs/Data-Void-Unsafe.html).

I just realized that @treeowl basically just made the same argument as I was writing this.

**Comment by treeowl** [comment]
Ah, but I think I was wrong about one point. Because it's parameterized by
mutability and not array type, there may (?) not be a problem restricting f
to take a representational argument....

**Comment by andrewthad** [comment]
@dolio I agree that `MutVar#` is much harder to deal with, but it's really only because people wouldn't want an immutable `Var#` that just acts as another pointer indirection. Interestingly, I don't seem to see `MutVar#` used much in large nested data structures (the kind of data structures that would benefit from the optimization). It's mostly arrays. Maybe one reason for this is that having a ton of `MutVar`s in a large nested data structure typically gives you abysmal GC performance since you end up having to scan all of them constantly (on every minor GC I think?).

@treeowl Hmmm... That'd be cool. If I follow correctly, restricting `f` to take a representational argument would mean this would be guaranteed to be sound, right?

**Comment by dolio** [comment]
Give `Immutable` and `Mutable` statuses like injective type families, so that you cannot match on them in a family. Similar to `Any`.

And yes, the point is that having an immutable `Var#` is undesirable. I forget whether things are actually set up in a way that makes this possible, because it's been too long since I spoke with Ed about it. But the idea is that when you freeze a mutable variable, you want it to be replaced by the sort of indirection that thunks use to point to things, which can then be fixed up by garbage collection to have no overhead (neither in representation, nor in API).

**Comment by andrewthad** [comment]
I've [written up some ideas](https://gist.github.com/andrewthad/4421a976ea05df7c41b126f948725acb) on a safe approach to deep freezing. It doesn't handle `MutVar#` at all but should work for all kinds of arrays.

**Comment by sjakobi** [comment]
What's the status of the ideas here? Is there anything viable with GHC 9.2 or 9.4?
